File: Algorithms/Sorting/QuickSort/QuickSort.js
/**
 * Pivot selected first element
 */
class QuickSort {
  sort(arr) {
    this.sortRecursive(arr, 0, arr.length - 1);
  }

  sortRecursive(arr, low, high) {
    if (low < high) {
      const pivotPosition = this.partition(arr, low, high);
      this.sortRecursive(arr, low, pivotPosition - 1);
      this.sortRecursive(arr, pivotPosition + 1, high);
    }
  }

  partition(arr, low, high) {
    let i = low,
      j = high;
    const pivot = arr[low];

    while (i < j) {
      while (arr.length - 1 != i && arr[i] <= pivot) i++;
      while (arr[j] > pivot) j--;

      if (i < j) [arr[i], arr[j]] = [arr[j], arr[i]];
    }
    [arr[j], arr[low]] = [arr[low], arr[j]];
    return j;
  }
}

File: Algorithms/Sorting/QuickSort/QuickSort.py
# Quicksort
# Select pivot first element of arr


class QuickSort():

    def sort(self, arr, low, high):
        if(low < high):
            j = self.partition(arr, low, high)
            self.sort(arr, low, j-1)
            self.sort(arr, j+1, high)

    def partition(self, arr, low, high):
        pivot = arr[low]
        i = low
        j = high
        while(i < j):
            while(len(arr) - 1 != i and arr[i] <= pivot):
                i += 1
            while(arr[j] > pivot):
                j -= 1

            if(i < j):
                arr[j], arr[i] = arr[i], arr[j]

        arr[j], arr[low] = pivot, arr[j]

        return j
